### ISSUE TITLE
Offline Latex support

### ISSUE THREAD

**Issue by entron** [uranusjr/macdown#358]
First of all, I love macdown very much and I use it everyday. Thanks for such a great tool!

When I sometime use the latex function, I noticed that by enabling the latex render in the setting, it makes macdown slow while typing and scrolling. This happens even though there are no latex currently on the visible area of the document. (I have a document about 10 A4 pages long, and it contains about 5 latex equations.) I guess this is because it constantly connect the internet for rendering the equation. Therefore I am wondering is that possible to make Latex available without internet? Maybe be downloading all the MathJax(?) to a local storage?

**Comment by uranusjr** [comment]
Serving MathJax locally is very difficult, and not recommended by the author (although it can be done). Many people (including myself) have considered switching to KaTeX instead (there’s even a pull request for that), but I don’t feel it is up to the requirements just yet. It works if all you want is simple equations, but as soon as you want moderately sophisticated `\text{}` blocks, `\array` environments, and many other things.

This is indeed an important issue. I will continue to monitor KaTeX, and see what we can do to improve it.

**Comment by entron** [comment]
KaTex looks promising! Is it possible to make it as one latex render option in the settings?

**Comment by RichardoC** [comment]
I personally already have TexLive as I use LaTeX quite a lot, can there be an option to use a local LaTeX distribution if ones already installed rather than needed an internet connection?

Thanks!

**Comment by franklinyu** [comment]
@RichardoC Compiling LaTeX to HTML [seems to be hard](http://tex.stackexchange.com/questions/39309/convert-latex-to-html).

**Comment by uranusjr** [comment]
@RichardoC @franklinyu Yeah, it is not really possible to go through LaTeX. Both MathJax and KaTeX are essentially re-implementation of the *syntax*; the code users write is the same, but the underlying parts are all different.

I do have a plan to integrate KaTeX directly into MacDown, removing Internet connection. But the implementation still lacks some things I consider important, so probably not yet.

**Comment by RichardoC** [comment]
Ah okay, thank you very much!

**Comment by jkulesza** [comment]
@RichardoC This is a very different response to your question/request, but have you tried LyX (www.lyx.org)?

**Comment by franklinyu** [comment]
@uranusjr I am not very optimistic about the future of KaTeX. By [Willie Wong](http://meta.math.stackexchange.com/questions/16809/a-mathjax-alternative-from-khan-academy#16817),

> I am also not entirely convinced that when feature parity is reached, KaTeX will still exhibit the same speed-up compared to MathJax, since both are based on a JavaScript backend.

So it remains to be seen.

**Comment by uranusjr** [comment]
@franklinyu The main reason I look into KaTeX is not speed (although it would be a very nice thing), but deployment. MathJax contains not less than 120 MB(!) worth of web fonts, which is more than 10 times larger than MacDown itself right now, and ridiculous for a simple editor. It is very possible most of them are not needed, but there’s absolutely not way to trace what can and what cannot be discarded.

KaTeX takes a better approach (I think), maintaining a custom set of fonts that contain only things that are absolutely necessary. It stands at at reasonable 2.5 MB right now, and the developers seem committed to keep it as compact as possible, which I feel comfortable with.

**Comment by RichardoC** [comment]
@jkulesza Thanks for the suggestion but if I'm doing something where I truly need LaTeX I tend to use TeXShop. I'm using MacDown when I just need to sketch out something (for example if I need some formulas typed up quickly)

**Comment by jkulesza** [comment]
@RichardoC Understood (though I still encourage you to give LyX a try ;-).  Also, for quick LaTeX-faithful formulas on OS X, you might check out LaTeXiT (https://www.chachatelier.fr/latexit/).

P.S. I'm not trying to drive anyone away from Macdown (I *really* like it); however, I'm just trying to spread knowledge about other LaTeX-oriented tools.

**Comment by franklinyu** [comment]
@uranusjr Yes, [Qute](http://www.inkcode.net/qute) (another Markdown editor) contains MathJax and the application is 100 MB after being extracted. It seems impossible to include MathJax while keeping below 20 MB.

KaTeX is not yet mature now, but I think we can start introducing the feature as an opt-in beta. Most people do not even need it, but for someone minimal offline rendering is better than nothing at all.

**Comment by kammoh** [comment]
Any updates on this issue? Has KaTeX matured enough to deserve another consideration?

**Comment by uranusjr** [comment]
No, but we’ve since then move MathJax hosting locally, so theoretically we can do this with MathJax as well. But that’s quite some work and I haven’t had time to work on it (and honestly wouldn’t in near future). Feel free to give it a shot if you’re interested.
